File: ios/Classes/service/ForegroundTask.swift
//
//  ForegroundTask.swift
//  flutter_foreground_task
//
//  Created by Woo Jin Hwang on 9/24/24.
//

import Foundation

private let BG_ISOLATE_NAME = "flutter_foreground_task/backgroundIsolate"
private let BG_CHANNEL_NAME = "flutter_foreground_task/background"

private let ACTION_TASK_START = "onStart"
private let ACTION_TASK_REPEAT_EVENT = "onRepeatEvent"
private let ACTION_TASK_DESTROY = "onDestroy"

class ForegroundTask {
  private let serviceStatus: BackgroundServiceStatus
  private let taskData: ForegroundTaskData
  private var taskEventAction: ForegroundTaskEventAction
  private let taskLifecycleListener: FlutterForegroundTaskLifecycleListener
  
  private var flutterEngine: FlutterEngine? = nil
  private var backgroundChannel: FlutterMethodChannel? = nil
  private var repeatTask: Timer? = nil
  private var isDestroyed: Bool = false
  
  init(
    serviceStatus: BackgroundServiceStatus,
    taskData: ForegroundTaskData,
    taskEventAction: ForegroundTaskEventAction,
    taskLifecycleListener: FlutterForegroundTaskLifecycleListener
  ) {
    self.serviceStatus = serviceStatus
    self.taskData = taskData
    self.taskEventAction = taskEventAction
    self.taskLifecycleListener = taskLifecycleListener
    initialize()
  }
  
  private func initialize() {
    guard let registerPlugins = SwiftFlutterForegroundTaskPlugin.registerPlugins else {
      print("Please register the registerPlugins function using the SwiftFlutterForegroundTaskPlugin.setPluginRegistrantCallback.")
      return
    }
    
    guard let callbackHandle = taskData.callbackHandle else {
      // no callback -> Unlike Android, the flutter engine does not start.
      return
    }
    
    // lookup callback
    let callbackInfo = FlutterCallbackCache.lookupCallbackInformation(callbackHandle)
    guard let entrypoint = callbackInfo?.callbackName else {
      print("Entrypoint not found in callback information.")
      return
    }
    guard let libraryURI = callbackInfo?.callbackLibraryPath else {
      print("LibraryURI not found in callback information.")
      return
    }
    
    // create flutter engine & execute callback
    let flutterEngine = FlutterEngine(name: BG_ISOLATE_NAME, project: nil, allowHeadlessExecution: true)
    let isRunningEngine = flutterEngine.run(withEntrypoint: entrypoint, libraryURI: libraryURI)
    
    if isRunningEngine {
      // register plugins
      registerPlugins(flutterEngine)
      taskLifecycleListener.onEngineCreate(flutterEngine: flutterEngine)
      
      // create background channel
      let messenger = flutterEngine.binaryMessenger
      let backgroundChannel = FlutterMethodChannel(name: BG_CHANNEL_NAME, binaryMessenger: messenger)
      backgroundChannel.setMethodCallHandler(onMethodCall)
      
      self.flutterEngine = flutterEngine
      self.backgroundChannel = backgroundChannel
    }
  }
  
  func onMethodCall(call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
      case "start":
        start()
      default:
        result(FlutterMethodNotImplemented)
    }
  }
  
  private func start() {
    runIfNotDestroyed {
      runIfCallbackHandleExists {
        let serviceAction = serviceStatus.action
        let starter: FlutterForegroundTaskStarter
        if serviceAction == .API_START || serviceAction == .API_RESTART || serviceAction == .API_UPDATE {
          starter = .DEVELOPER
        } else {
          starter = .SYSTEM
        }
        
        backgroundChannel?.invokeMethod(ACTION_TASK_START, arguments: starter.rawValue) { _ in
          self.runIfNotDestroyed {
            self.startRepeatTask()
          }
        }
        taskLifecycleListener.onTaskStart(starter: starter)
      }
    }
  }
  
  private func invokeTaskRepeatEvent() {
    backgroundChannel?.invokeMethod(ACTION_TASK_REPEAT_EVENT, arguments: nil)
    taskLifecycleListener.onTaskRepeatEvent()
  }
  
  private func startRepeatTask() {
    stopRepeatTask()
    
    let type = taskEventAction.type
    let interval = TimeInterval(Double(taskEventAction.interval) / 1000)
    
    if type == .NOTHING {
      return
    }
    
    if type == .ONCE {
      invokeTaskRepeatEvent()
      return
    }
    
    repeatTask = Timer.scheduledTimer(withTimeInterval: interval, repeats: true) { _ in
      self.invokeTaskRepeatEvent()
    }
  }
  
  private func stopRepeatTask() {
    repeatTask?.invalidate()
    repeatTask = nil
  }
  
  func invokeMethod(_ method: String, arguments: Any?) {
    runIfNotDestroyed {
      backgroundChannel?.invokeMethod(method, arguments: arguments)
    }
  }
  
  func update(taskEventAction: ForegroundTaskEventAction) {
    runIfNotDestroyed {
      runIfCallbackHandleExists {
        self.taskEventAction = taskEventAction
        startRepeatTask()
      }
    }
  }
  
  func destroy() {
    runIfNotDestroyed {
      stopRepeatTask()
      
      backgroundChannel?.setMethodCallHandler(nil)
      if taskData.callbackHandle == nil {
        taskLifecycleListener.onEngineWillDestroy()
        flutterEngine?.destroyContext()
        flutterEngine = nil
      } else {
        backgroundChannel?.invokeMethod(ACTION_TASK_DESTROY, arguments: nil) { _ in
          self.flutterEngine?.destroyContext()
          self.flutterEngine = nil
        }
        taskLifecycleListener.onTaskDestroy()
        taskLifecycleListener.onEngineWillDestroy()
      }
      
      isDestroyed = true
    }
  }
  
  private func runIfCallbackHandleExists(call: () -> Void) {
    if taskData.callbackHandle == nil {
      return
    }
    call()
  }

  private func runIfNotDestroyed(call: () -> Void) {
    if isDestroyed {
      return
    }
    call()
  }
}

File: android/src/main/kotlin/com/pravera/flutter_foreground_task/service/ForegroundTask.kt
package com.pravera.flutter_foreground_task.service

import android.content.Context
import android.util.Log
import com.pravera.flutter_foreground_task.FlutterForegroundTaskLifecycleListener
import com.pravera.flutter_foreground_task.FlutterForegroundTaskStarter
import com.pravera.flutter_foreground_task.models.ForegroundServiceAction
import com.pravera.flutter_foreground_task.models.ForegroundServiceStatus
import com.pravera.flutter_foreground_task.models.ForegroundTaskData
import com.pravera.flutter_foreground_task.models.ForegroundTaskEventAction
import com.pravera.flutter_foreground_task.models.ForegroundTaskEventType
import io.flutter.FlutterInjector
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.embedding.engine.dart.DartExecutor
import io.flutter.embedding.engine.loader.FlutterLoader
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.view.FlutterCallbackInformation
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class ForegroundTask(
    context: Context,
    private val serviceStatus: ForegroundServiceStatus,
    private val taskData: ForegroundTaskData,
    private var taskEventAction: ForegroundTaskEventAction,
    private val taskLifecycleListener: FlutterForegroundTaskLifecycleListener,
) : MethodChannel.MethodCallHandler {
    companion object {
        private val TAG = ForegroundTask::class.java.simpleName

        private const val ACTION_TASK_START = "onStart"
        private const val ACTION_TASK_REPEAT_EVENT = "onRepeatEvent"
        private const val ACTION_TASK_DESTROY = "onDestroy"
    }

    private val flutterEngine: FlutterEngine
    private val flutterLoader: FlutterLoader
    private val backgroundChannel: MethodChannel
    private var repeatTask: Job? = null
    private var isDestroyed: Boolean = false

    init {
        // create flutter engine
        flutterEngine = FlutterEngine(context)
        flutterLoader = FlutterInjector.instance().flutterLoader()
        if (!flutterLoader.initialized()) {
            flutterLoader.startInitialization(context)
        }
        flutterLoader.ensureInitializationComplete(context, null)
        taskLifecycleListener.onEngineCreate(flutterEngine)

        // create background channel
        val messenger = flutterEngine.dartExecutor.binaryMessenger
        backgroundChannel = MethodChannel(messenger, "flutter_foreground_task/background")
        backgroundChannel.setMethodCallHandler(this)

        // execute callback
        val callbackHandle = taskData.callbackHandle
        if (callbackHandle != null) {
            val bundlePath = flutterLoader.findAppBundlePath()
            val callbackInfo = FlutterCallbackInformation.lookupCallbackInformation(callbackHandle)
            val dartCallback = DartExecutor.DartCallback(context.assets, bundlePath, callbackInfo)
            flutterEngine.dartExecutor.executeDartCallback(dartCallback)
        }
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "start" -> start()
            else -> result.notImplemented()
        }
    }

    private fun start() {
        runIfNotDestroyed {
            runIfCallbackHandleExists {
                val serviceAction = serviceStatus.action
                val starter = if (serviceAction == ForegroundServiceAction.API_START ||
                    serviceAction == ForegroundServiceAction.API_RESTART ||
                    serviceAction == ForegroundServiceAction.API_UPDATE) {
                    FlutterForegroundTaskStarter.DEVELOPER
                } else {
                    FlutterForegroundTaskStarter.SYSTEM
                }

                backgroundChannel.invokeMethod(ACTION_TASK_START, starter.ordinal) {
                    runIfNotDestroyed {
                        startRepeatTask()
                    }
                }
                taskLifecycleListener.onTaskStart(starter)
            }
        }
    }

    private fun invokeTaskRepeatEvent() {
        backgroundChannel.invokeMethod(ACTION_TASK_REPEAT_EVENT, null)
        taskLifecycleListener.onTaskRepeatEvent()
    }

    private fun startRepeatTask() {
        stopRepeatTask()

        val type = taskEventAction.type
        val interval = taskEventAction.interval

        if (type == ForegroundTaskEventType.NOTHING) {
            return
        }

        if (type == ForegroundTaskEventType.ONCE) {
            invokeTaskRepeatEvent()
            return
        }

        repeatTask = CoroutineScope(Dispatchers.Default).launch {
            while (true) {
                delay(interval)
                withContext(Dispatchers.Main) {
                    try {
                        invokeTaskRepeatEvent()
                    } catch (e: Exception) {
                        Log.e(TAG, "repeatTask", e)
                    }
                }
            }
        }
    }

    private fun stopRepeatTask() {
        repeatTask?.cancel()
        repeatTask = null
    }

    fun invokeMethod(method: String, data: Any?) {
        runIfNotDestroyed {
            backgroundChannel.invokeMethod(method, data)
        }
    }

    fun update(taskEventAction: ForegroundTaskEventAction) {
        runIfNotDestroyed {
            runIfCallbackHandleExists {
                this.taskEventAction = taskEventAction
                startRepeatTask()
            }
        }
    }

    fun destroy() {
        runIfNotDestroyed {
            stopRepeatTask()

            backgroundChannel.setMethodCallHandler(null)
            if (taskData.callbackHandle == null) {
                taskLifecycleListener.onEngineWillDestroy()
                flutterEngine.destroy()
            } else {
                backgroundChannel.invokeMethod(ACTION_TASK_DESTROY, null) {
                    flutterEngine.destroy()
                }
                taskLifecycleListener.onTaskDestroy()
                taskLifecycleListener.onEngineWillDestroy()
            }

            isDestroyed = true
        }
    }

    private fun runIfCallbackHandleExists(call: () -> Unit) {
        if (taskData.callbackHandle == null) {
            return
        }
        call()
    }

    private fun runIfNotDestroyed(call: () -> Unit) {
        if (isDestroyed) {
            return
        }
        call()
    }

    private fun MethodChannel.invokeMethod(method: String, data: Any?, onComplete: () -> Unit = {}) {
        val callback = object : MethodChannel.Result {
            override fun success(result: Any?) {
                onComplete()
            }

            override fun error(errorCode: String, errorMessage: String?, errorDetails: Any?) {
                onComplete()
            }

            override fun notImplemented() {
                onComplete()
            }
        }
        invokeMethod(method, data, callback)
    }
}
